producer/awsS3.go

```diff
@@ -0,0 +1,281 @@
+
+		prod.filesByStream[streamID] = batchedFile
+		prod.files[baseFileName] = batchedFile
+	}
+
+	// Close existing batchedFile.writer
+	if batchedFile.HasWriter() {
+		oldAwsWriter := batchedFile.GetWriterAndUnset()
+
+		prod.Logger.Info("Rotated ", oldAwsWriter.Name(), " -> ", baseFileName)
+		go oldAwsWriter.Close() // close in subroutine for eventually compression in the background
+	}
+
+	// Update BatchedWriterAssembly writer
+	writer := awsS3.NewBatchedFileWriter(prod.s3Client, prod.bucket, prod.getFinalFileName(baseFileName), prod.Logger)
+	batchedFile.SetWriter(&writer)
+
+	return batchedFile, nil
+}
+
+func (prod *AwsS3) needsRotate(batchedFile *components.BatchedWriterAssembly, forceRotate bool) (bool, error) {
+	// run default rotation checks
+	if needUpload, err := batchedFile.NeedsRotate(prod.Rotate, forceRotate); needUpload {
+		return true, err
+	}
+
+	// check if max multipart uploads of 1000 reached
+	// @see: http://docs.aws.amazon.com/AmazonS3/latest/dev/mpuoverview.html
+	// we use 995 to have a small buffer to the limit and need at least +1 upload part for the last flush
+	writer, ok := batchedFile.GetWriter().(awsS3.BatchedFileWriterInterface)
+	if ok && writer.GetUploadCount() > 995 {
+		prod.Logger.Debug("Rotate true: ", "upload count reached limit of 1000")
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (prod *AwsS3) getBaseFileName(streamID core.MessageStreamID) string {
+	if prod.hasWildcard {
+		streamName := core.StreamRegistry.GetStreamName(streamID)
+		return strings.Replace(prod.fileNamePattern, "*", streamName, -1)
+	}
+
+	return prod.fileNamePattern
+}
+
+func (prod *AwsS3) getFinalFileName(baseFileName string) string {
+	fileExt := filepath.Ext(baseFileName)
+	fileName := baseFileName[:len(baseFileName)-len(fileExt)]
+
+	timestamp := time.Now().Format(prod.Rotate.Timestamp)
+	signature := fmt.Sprintf("%s_%s", fileName, timestamp)
+	maxSuffix := uint64(0)
+
+	//todo: get list from aws?
+	/*files, _ := ioutil.ReadDir(streamFile.dir)
+	for _, f := range files {
+		if strings.HasPrefix(f.Name(), signature) {
+			// Special case.
+			// If there is no extension, counter stays at 0
+			// If there is an extension (and no count), parsing the "." will yield a counter of 0
+			// If there is a count, parsing it will work as intended
+			counter := uint64(0)
+			if len(f.Name()) > len(signature) {
+				counter, _ = tstrings.Btoi([]byte(f.Name()[len(signature)+1:]))
+			}
+
+			if maxSuffix <= counter {
+				maxSuffix = counter + 1
+			}
+		}
+	}*/
+
+	if maxSuffix == 0 {
+		return fmt.Sprintf("%s%s", signature, fileExt)
+	}
+
+	formatString := "%s_%d%s"
+	if prod.Rotate.ZeroPad > 0 {
+		formatString = fmt.Sprintf("%%s_%%0%dd%%s", prod.Rotate.ZeroPad)
+	}
+	return fmt.Sprintf(formatString, signature, int(maxSuffix), fileExt)
+
+}
+
+func (prod *AwsS3) newBatchedWriterAssembly() *components.BatchedWriterAssembly {
+	batchedFile := components.NewBatchedWriterAssembly(
```

Comment:
Move this function to the BatchedFileBuilder type as suggested above.
So you can e.g. call `builder.New(prod)` to get a new batchedFile.